### ISSUE TITLE
Configuration for multiple projects

### ISSUE THREAD

**Issue by stefpiatek** [UCLH-Foundry/PIXL#232]
### Definition of Done / Acceptance Criteria

PIXL pipeline should have configuration stored in files or database to be able to run the pipeline for multiple projects:

- [x] DICOM tag actions per project. Pointing to files (keeping 100 days and #284 being committed to the repo so we can use PRs)
- [x] whether images should be send via DICOMWeb or FTPS
- [x] FTPS connection details for project
- [ ] DICOMWeb connection details for sending data to endpoint

### Testing

No need to test DICOMWeb configuration at this stage, but will do this once we're sending via DICOMWeb API

### Documentation

Update documentation for how configuration gets set 

### Details and Comments

Currently we have single values set in the `.env` file or hardcoded details, but the PIXL pipeline should be able to be reused for multiple projects without any changes to the code

**Comment by stefpiatek** [comment]
@timroberts would like to participate with this as there are touchpoints with OMOP ES, and @HChughtai with MIRSG & IoN DRI

**Comment by milanmlft** [comment]
>to run the pipeline for multiple projects

Meaning we want to be able to do

```sh
pixl populate /resources/omop
pixl start
...
```

and have the whole pipeline run for multiple projects?

Or rather something like

```sh
pixl populate /resources/omop

# project 1
pixl start --config .project1_config

# project 2
pixl start --config .project2_config

# etc.
```

**Comment by milanmlft** [comment]
As a side note, [Apple has released a new "configuration" programming language called Pkl]([Introducing Pkl, a programming language for configuration :: Pkl Docs](https://pkl-lang.org/blog/introducing-pkl.html)) which actually might be a good use case here. Though it's quite new, so not sure how stable it is 😬

**Comment by stefpiatek** [comment]
> As a side note, [Apple has released a new "configuration" programming language called Pkl](https://pkl-lang.org/blog/introducing-pkl.html) which actually might be a good use case here. Though it's quite new, so not sure how stable it is 😬

That IS fun, but no python bindings. [dynaconf](https://www.dynaconf.com/) is a pretty nice way to deal with secrets in python

**Comment by peshence** [comment]
Questions:
- The VOXL diagram implies parquet always moves to FTPS and images always move to DICOMWeb instances. Is that the plan?
- similar question - assuming FTPS is still an option for images, is the FTPS setting the same for parquet and for images? IE they are different per project
- Azure storage - I see we currently offload the whole anon db table there, but have not seen anything else connecting to it (except I guess a DicomWeb Azure instance but that shouldn't be a separate case?). 

Notes:
- ![this line](https://github.com/UCLH-Foundry/PIXL/blob/92555e6e6e887947ad4b8fa13bfb189ee69ee933/orthanc/orthanc-anon/plugin/pixl.py#L266) refuses all but XRay modalities which presumably we will also need to edit
- As it stands, I think we can achieve most of what is needed by passing the config to the Orthanc Anon instance (by perhaps adding DICOM tags? don't immediately see another way). I think the upload options would make more sense once we have a singlular upload service (I heard talk of this but not sure if it's the plan?)

**Comment by stefpiatek** [comment]
> The VOXL diagram implies parquet always moves to FTPS and images always move to DICOMWeb instances. Is that the plan?

Nah Images can go to azure, DicomWeb and FTPS which I think is what's in the voxl diagram

>similar question - assuming FTPS is still an option for images, is the FTPS setting the same for parquet and for images? IE they are different per project

If a project uses FTPS for both imaging and parquet, then they'd use the same connection details. But you could have imaging using one modality and parquet using FTPS

> Azure storage - I see we currently offload the whole anon db table there, but have not seen anything else connecting to it (except I guess a DicomWeb Azure instance but that shouldn't be a separate case?).

Yep, that's a possible future for azure blob storage, not current. 


> As it stands, I think we can achieve most of what is needed by passing the config to the Orthanc Anon instance (by perhaps adding DICOM tags? don't immediately see another way). I think the upload options would make more sense once we have a singlular upload service (I heard talk of this but not sure if it's the plan?)

I think that's what I expected but depends what you mean by passing config. From the VOXL plan I thought: 

- We store the project name in the pixl database along with the image identfier and pseudonymised id. 
- In raw we add the project name to the dicom tags,
- in anon the project name lets us find the bound `projects` dir so we can read the config and know what anonyisation needs to be applied. We also query the pixl database using the project name and image identifier, to find the pseudonymised id and do the anonymisation .
-  After the anon image is stable we then work out where to send it using the pseudonymised id. We could do this by requering or building up a local cache that maps anonymisation ids to projects. 

Is that what you had in mind?

**Comment by peshence** [comment]
> Nah Images can go to azure, DicomWeb and FTPS which I think is what's in the voxl diagram
Ah you're right, I missed one arrow.

> Yep, that's a possible future for azure blob storage, not current.
I guess we ignore this option for now then - it's only dicomweb or sftp? Could we want to upload ot more than one place?
>     * In raw we add the project name to the dicom tags,
Not sure why don't we just do this in anon if we're querying the db anyway?

> Is that what you had in mind?
No but this is way better 😅 

I am a bit overwhelmed with the details, so let me rephrase it slightly to understand better:
- load input files. Write project name, and private/public ids to db
- later in orthanc, use id which already exists in dicom tags to query db and get which project it belongs to
- use that to get appropriate config?

**Comment by stefpiatek** [comment]
> Not sure why don't we just do this in anon if we're querying the db anyway?

Ah yes sorry, that part was muddled. It'll be imaging-api that adds this to orthanc raw (because it has the entire message from the queue) after the orthanc raw job has completed (this will happen if the image already exists in orthanc raw, so if an image is reused across multiple studies it'll get routed to the right place in anonymisation). So it wont require the database at that point. I think that should work out nicely anyway

> I am a bit overwhelmed with the details, so let me rephrase it slightly to understand better:
> 
>     load input files. Write project name, and private/public ids to db
>     later in orthanc, use id which already exists in dicom tags to query db and get which project it belongs to
>     use that to get appropriate config?
> 

Yeah that sounds right, depending on where we get the project name from

**Comment by stefpiatek** [comment]
will be using template similar to cookiecutter and test out with projects

**Comment by peshence** [comment]
Decided to tackle this from two sides and converge later:
# @peshence :
Get project name from input parquet to orthanc-anon/export-api by either
### DB
- [ ] load project name to db in cli/populate
- [ ] query db and match said project name to dicom image in orthanc-anon
### DICOM tags
- [ ] alternatively add project name to dicom tags in imaging-api @stefpiatek if I got that right?
# @milanmlft :
- [ ] generate config example and template
- [ ] add config loading functionality (probably as part of core, to be used in orthanc-anon and export api)

**Comment by stefpiatek** [comment]
>  load project name to db in cli/populate

You'll be pleased to hear that this is already done

**Comment by peshence** [comment]
Questions:
- is there a reason some methods in orthanc-anon use logger, but some use orthanc? I assume no and we should switch all to orthanc, but have not done so.
- will we ever send via STOW? is that the eventual send-to-dicomweb?

**Comment by stefpiatek** [comment]
> is there a reason some methods in orthanc-anon use logger, but some use orthanc? I assume no and we should switch all to orthanc, but have not done so.

I don't think so, though they do come up differently when running

> will we ever send via STOW? is that the eventual send-to-dicomweb?

Yeah we want to keep STOW as an option for that

I've defined test--ftp--host ,test--ftp--username ,test--ftp--password, test--ftp--port  in the keyvault (setting to whats used in the test .env while you're getting access to it. Keyvault name is kv-pixl-dev

**Comment by stefpiatek** [comment]
FYI project secrets added to github 
<img width="290" alt="image" src="https://github.com/UCLH-Foundry/PIXL/assets/8124189/35b123c3-18cb-4ee8-a83a-e46d6b79602f">

**Comment by milanmlft** [comment]
>FYI project secrets added to github

What's the `EXPORT_AZ_NAME` for? 🤔 
I can fetch stuff from the Azure KeyVault with only he 4 other variables.

**Comment by stefpiatek** [comment]
> What's the `EXPORT_AZ_NAME` for? 🤔

Ha probably not needed then, ignore it!
